jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c

```diff
@@ -1276,6 +1276,220 @@ ecma_builtin_typedarray_prototype_fill (ecma_value_t this_arg, /**< this argumen
+        {
+          // Keep NaNs at the end of the array.
+          result = ECMA_NUMBER_MINUS_ONE;
+        }
+        else if (lhs < rhs)
+        {
+          result = ECMA_NUMBER_MINUS_ONE;
+        }
+        else if (lhs > rhs)
+        {
+          result = ECMA_NUMBER_ONE;
+        }
+        else
+        {
+          result = ECMA_NUMBER_ZERO;
+        }
+      }
+      else
+      {
+        /*
+         * comparefn, if not undefined, will always contain a callable function object.
+         * We checked this previously, before this function was called.
+         */
+        JERRY_ASSERT (ecma_op_is_callable (comparefn));
+        ecma_object_t *comparefn_obj_p = ecma_get_object_from_value (comparefn);
+
+        ecma_value_t compare_args[] = {j, k};
+
+        ECMA_TRY_CATCH (call_value,
+                        ecma_op_function_call (comparefn_obj_p,
+                                               ECMA_VALUE_UNDEFINED,
+                                               compare_args,
+                                               2),
+                        ret_value);
+
+        if (!ecma_is_value_number (call_value))
+        {
+          ECMA_OP_TO_NUMBER_TRY_CATCH (ret_num, call_value, ret_value);
+          result = ret_num;
+          ECMA_OP_TO_NUMBER_FINALIZE (ret_num);
+
+          // If the coerced value can't be represented as a Number, compare them as equals.
+          if (ecma_number_is_nan (result))
+          {
+            result = ECMA_NUMBER_ZERO;
+          }
+        }
+        else
+        {
+          result = ecma_get_number_from_value (call_value);
+        }
+
+        ECMA_FINALIZE (call_value);
+      }
+    }
+  }
+
+  if (ecma_is_value_empty (ret_value))
+  {
+    ret_value = ecma_make_number_value (result);
+  }
+
+  return ret_value;
+} /* ecma_builtin_typedarray_prototype_sort_compare_helper */
+
+/**
+ * The %TypedArray%.prototype object's 'sort' routine.
+ *
+ * See also:
+ *          ES2015, 22.2.3.25, 22.1.3.24
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_typedarray_prototype_sort (ecma_value_t this_arg, /**< this argument */
+                                        ecma_value_t comparefn) /**< comparator fn */
+{
+  if (!ecma_is_typedarray (this_arg))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is not a TypedArray."));
+  }
+
+  if (!ecma_is_value_undefined (comparefn) && !ecma_op_is_callable (comparefn))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Compare function is not callable."));
+  }
+
+  ecma_object_t *typedarray_p = ecma_get_object_from_value (this_arg);
+  ecma_length_t typedarray_length = ecma_typedarray_get_length (typedarray_p);
+
+  if (!typedarray_length)
+  {
+    return ecma_copy_value (this_arg);
+  }
+
+  ecma_value_t ret_value = ECMA_VALUE_EMPTY;
+
+  JMEM_DEFINE_LOCAL_ARRAY (values_buffer, typedarray_length, ecma_value_t);
```

Comment:
I don't think it would work if I just gave the heap-sort helper an underlying buffer without having to change the implementation -- then that means my helper isn't generic anymore. The heap-sort seems to work on `ecma_value_t`s and I don't think I can just map variable-sized (ie. 8-bit integers, 32-bit floats, etc.) arraybuffer elements to it?

Reply:
Ok, at least this is good for binary size. We can optimize the code later if needed, since heapsort is not exactly a complicated algorithm.